### ISSUE TITLE
Offline and invisible mode - feature request

### ISSUE THREAD

**Issue by ghost** [qTox/qTox#2602]
You can change your status to "Online", "Away", and "Busy", but I think that you should also be able to change it to "Offline" and "Invisible". "Offline" being when you are offline though you can read your messages etc (but can't send or receive any). And "Invisible" being when you appear offline to others but are in fact just invisible.

**Comment by ProMcTagonist** [comment]
https://github.com/tux3/qTox/issues/826

**Comment by zetok** [comment]
`Invisible` is not doable - as for `offline`, why not.

**Comment by ghost** [comment]
Ok, just out of interest, why is `Invisible` not doable?

**Comment by Spencer-H** [comment]
Tox is a p2p network. You must tell your contacts when you are online in able to be able to send and receive messages. If you go 'invisible' and stop telling your friends that you are online, they will no longer be able to connect to you and thus wont send messages, making you actually offline.

**Comment by ghost** [comment]
@Spencer-H @zetok: Yes, but it doesn't have to quite work like that. The qTox client will still tell other clients that it is online, however it will tell them to display it as though it is not online. So when it is "Online", "Away", or "Busy", it will transmit that message and that is what people will see, and when it is "Invisible" it will do the exact same thing except it will have it so that all the other clients display the "Offline" symbol for them. So the client tells the others that it is online but to pretend and tell the user that it is not.

I mean, if the user looks very far into it they may be able to see that the person is actually online, but that doesn't matter so much.

**Comment by Spencer-H** [comment]
Tox as a whole assumes that there is as little trust as possible between clients. And because of the way clients are implemented, you would then have a case where you would only be offline to people who had clients that would allow this. For example, qtox might implement it, but utox might not, and then people on utox would be able to see you as online anyway and you'd just be making your friends on qTox angry.

**Comment by ghost** [comment]
You might make them angry, so then it should warn you that "Invisible" mode would only show you as "Offline" to those who have the clients which have implemented this. I think that there should at least be the optional ability to have this option if you also give the warning. So perhaps one by default wouldn't be able to change their status to "Invisible", but would have to go into the settings where the warning is also located in order to make it so that they have that option available.

**Comment by Spencer-H** [comment]
But who would leave their client turned on to accept invisible? It'd be just as effective as messaging your friends and saying "Please don't message me for a while", or just setting yourself to busy, which DOES display to your friends.

**Comment by ghost** [comment]
Well, some people may not want their friends to know when they are online. Or they may not know the people in person that they have on their contacts list and may for some reason not always have them know whether they are online or not. I mean, I don't really use the feature any more on most messaging apps, but a lot of people do and it would be a good at least optional feature to give people lots of options.

**Comment by Spencer-H** [comment]
The problem is that it's an option that requires the cooperation of those
users. You're technically just asking them to treat you as offline each
time, nothing is forcing them to do som and it's a bad idea to be able to
FORCE people to see someone as offline when they're not so. A more
effective method would be to add the option to 'hide' from select people,
which would just stop your client connecting or responding to them for a
while. I don't know how this would work in the DHT but it may be
possible... Don't know if you should get your hopes up for it though.

**Comment by zetok** [comment]
@SeventhSonOfASeventhSon 

> The qTox client will still tell other clients that it is online, however it will tell them to display it as though it is not online.

Right.

So, what @Spencer-H wrote is accurate from higher-level overview. When it comes down to how things actually work, qTox doesn't have much to say in that matter - it's all "managed" by toxcore, while qTox only has the choice between "online", "away" and "busy".

With that being said (if nothing has changed) toxcore is not implementing "invisible" status, thus it's not doable. At least not in a proper way, which is the only sane way of doing things, if one expects them to actually work.

Personally, I don't see point in "invisible" status.. not when it's just up to client whether they decide to ignore "invisibility" or not. And when it comes to that, as @Spencer-H wrote, there's no other way of how things could work.

# 

> A more effective method would be to add the option to 'hide' from select people, which would just stop your client connecting or responding to them for a while.

That's how "blocking" contacts would be done :D
I.e. "blocked" contacts would be removed from "friends", as toxcore sees it, thus it will ~not connect to them.

The way I see it, "blocking" would mostly mitigate need for "invisible" status, with some minor exceptions to that.

**Comment by shalnoff** [comment]
I'm supporting the opinion of @SeventhSonOfASeventhSon regarding the personal Blocking feature. It would be great, with ability to add additional comment (like "I'm busy, sorry") or just without.

Nevertheless, as far as invisible option couldn't be implemented directly by design, it may be created via multiple accounts support, IMHO. I.e. an user has separate accounts (UIDs) for different groups (say: business, relatives, friends) and, consequently, could just switch one off when others leave online. 

I'll try to add it as a proposal issue.

**Comment by shalnoff** [comment]
BTW the solution was offered in another tread (Multipe UIDs), I just move it here, in relevant context. 

This feature quite easily could be done without changing anything in toxcore, just by making user removed and all of his messages ignored. I.e. qTox leave the record of user but mark it as "ignored" and remove it for Tox network. Consequently: if the user decide do get another user back, he just restore the previous status and qTox add the same user to the list for Tox Core . 

The only doubtful issue is an implementation of New User request mechanism in Tox Core. Does it have obligatory requesting on each new inquiry? is it possible to add UID "silently"?

**Comment by TheNain38** [comment]
No, there is the `friendAddNoRequest` function which adds, a friend but without sending a friend request, and it only needs the PublicKet, no noSpam

**Comment by ProMcTagonist** [comment]
#826

**Comment by UralZima** [comment]
Please support toxcore issue: https://github.com/TokTok/c-toxcore/issues/522

**Comment by anthonybilinski** [comment]
Using toxcore's `tox_friend_add_norequest` API should work to accomplish contact blocking. More context here: https://github.com/qTox/qTox/issues/2909. It seems like a good idea to me and doesn't depend on any changes in toxcore.
